lib/build.ts

```diff
@@ -187,3 +194,169 @@ export async function runBuildTask(
+		try {
+			const manifest = await getManifest(docker.modem, repository);
+			return manifest.Descriptor.mediaType === MEDIATYPE_MANIFEST_V1;
+		} catch {
+			// eat the exception... yummy!
+		}
+
+		// at this time, treat any errors as true (v1 manifest)
+		return true;
+	};
+
+	if (task.imageName) {
+		imageReferences = [task.imageName + !!task.tag ? `:${task.tag}` : ''];
+	} else {
+		if (!task.dockerfile) {
+			// Sanity check
+			console.warn('Build task does not have an associated Dockerfile');
+			return false;
+		}
+		const { DockerfileParser } = await import('dockerfile-ast');
+		const parsedDockerfile = DockerfileParser.parse(task.dockerfile);
+		const dockerInstructions = parsedDockerfile.getInstructions();
+		const fromInstructions = dockerInstructions.filter(
+			(inst) => inst.getKeyword() === 'FROM',
+		);
+		if (fromInstructions.length === 0) {
+			// Sanity check
+			console.warn(
+				'Build task associated Dockerfile does no reference any images',
+			);
+			return false;
+		}
+		imageReferences = [];
+		fromInstructions.forEach((inst) => {
+			const fromLineArguments = inst.getArguments();
+			if (fromLineArguments.length === 0) {
+				// bail out, no image specified?
+				return;
+			}
+
+			imageReferences.push(fromLineArguments[0].getValue());
+		});
+	}
+
+	const outdatedBalenalibImages: string[] = [];
+	const v1Images: string[] = [];
+	const v2Images: string[] = [];
+
+	await Promise.all(
+		imageReferences.map(async (r) => {
+			const hasV1Manifest = await checkHasV1Manifest(r);
+			if (r.indexOf('balenalib/') >= 0) {
+				if (hasV1Manifest) {
+					outdatedBalenalibImages.push(r);
+				}
+			} else if (hasV1Manifest) {
+				v1Images.push(r);
+			} else {
+				v2Images.push(r);
+			}
+		}),
+	);
+
+	if (outdatedBalenalibImages.length > 0) {
+		throw new BuildProcessError(
+			'Error: The following balenalib images do not support platform targeting. Please update them:\n-' +
+				outdatedBalenalibImages.join('\n-'),
+		);
+	}
+	if (v1Images.length === 0) {
+		// Only v2 images, let Docker receive `--platform`
+		return true;
+	} else if (v2Images.length > 0) {
+		// Mixed v1 and v2 images
+		console.warn(
+			'Warning: Dockerfile contains a mixes images that require an architecture selection with ones that do not support it.\n' +
+				'The following images do not support architecture selection:\n-' +
+				v1Images.join('\n-') +
+				'\n\n' +
+				'The following images require architecture selection:\n-' +
+				v2Images.join('\n-') +
+				'\n\n' +
+				'To avoid an image build error, the CLI has disabled architecture selection. ' +
+				'However, this means that the architecture of the machine where the Docker ' +
+				'Engine is running will be used by default to select the architecture of the ' +
+				'images that require architecture selection, which may result in "exec format ' +
+				'error" at runtime. To avoid this problem, please update or replace the images ' +
```

Comment:
```suggestion
				'error" at runtime. To avoid this warning, please update or replace the images ' +
```
It's more likely they want the warning to go away, as they might not consider it to be a "problem" even if we do.